### PR TITLE
Pivot table calculation and the GetPivotData function support

### PR DESCRIPTION
**Pull request by JanKallman**
## Pivot Data Calculation
Added functions & propeties
* ExcelPivotTable.Calculate method
* ExcelPivotTable.GetPivotData method
* ExcelPivotTable.IsCalculated Property
### Calculating Data Fields on rows and columns:
Functions:
* Sum
* Count
* Average
* Min
* Max
* Product
* Count Numbers
* StdDev
* StdDevP 
* Var
* VarP

### Support for Show Value as 
Show Value As:
* % of Grand Total
* % of Column Total
* % of Row Total
* % Of
* % of Parent Row Total
* % of Parent Column Total
* % of Parent Total
*  Difference From
*  % Difference From
*  Running Total in
* % Running Total in
* Rank smallest to largest
* Rank largest to smallest
* Index

### Filters

Caption filters
Numeric Filters
Date & Time Filters
Value Filters
Top/bottom filters
Slicers
### Current limitations:
* EPPlus will not handle external data sources, data models or OLAP data sources.

### CODE REVIEW

File: src/EPPlus/FormulaParsing/DependencyChain/RpnFormulaExecution.cs [hunk 1]
```diff
@@ -792,15 +799,20 @@ private static FormulaRangeAddress ExecuteNextToken(RpnOptimizedDependencyChain
                         s.Push(s.Pop().Negate());
                         break;
                     case TokenType.CellAddress:
-                    case TokenType.ExcelAddress:                    
-                        var e = f._expressions[f._tokenIndex];
+                    case TokenType.ExcelAddress:
+						var e = f._expressions[f._tokenIndex];
```
Comment (OssianEPPlus) [inline src/EPPlus/FormulaParsing/DependencyChain/RpnFormulaExecution.cs:803]: Indentation seems to have gotten a bit weird in this switch case

File: src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+            return new CompileResult(result, DataType.Decimal);
+        }
+
+        private void AddCriterias(List<string> fieldValues, ExcelPivotTableRowColumnFieldCollection fields, ref List<PivotDataFieldItemSelection> criteria)
+        {
+            for (int i=0;i<fieldValues.Count-1; i++)
+            {
+                criteria.Add(new PivotDataFieldItemSelection(fields[i].Name, fieldValues[i]));
+            }
+            criteria.Add(new PivotDataFieldItemSelection(fields[fieldValues.Count - 1].Name, null));
```
Comment (OssianEPPlus) [inline src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs:184]: Why does the last critera not get the fieldValue?

File: src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+
+        private bool GetSubTotalFunctionFromString(string value, out eSubTotalFunctions function)
+        {
+            switch (value.ToLower())
+            {
+                case "sum":
+                    function = eSubTotalFunctions.Sum;
+                    break;
+                case "count":
+                    function = eSubTotalFunctions.CountA;
```
Comment (OssianEPPlus) [inline src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs:224]: We return CountA for `count` case. Yet the enum comment for count a is Count, Numbers. and `count nums` case returns Count. 

Are the two mixed up in the switch case or is the comment wrong?

File: src/EPPlus/Style/ExcelNumberFormat.cs [hunk 4]
```diff
@@ -201,5 +204,92 @@ internal static int GetFromBuildIdFromFormat(string format)
                     return int.MinValue;
             }
         }
-    }
+        internal NumberFormatType _numberformatType= NumberFormatType.Unset;
+
+		internal bool IsDateFormat
```
Comment (OssianEPPlus) [inline src/EPPlus/Style/ExcelNumberFormat.cs:209]: IsDateFormat and IsNumberFormat appear to have no references

File: src/EPPlus/Table/PivotTable/Calculation/ArrayComparer.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+namespace OfficeOpenXml.Table.PivotTable
+{
+    internal partial class PivotTableCalculation
+    {
+		internal static PivotCalculationStore GetNewCalculatedItems()
+		{
+			return new PivotCalculationStore();
+		}
+		internal static Dictionary<int[], HashSet<int[]>> GetNewKeys()
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/Calculation/ArrayComparer.cs:25]: No references to GetNewKeys

File: src/EPPlus/Table/PivotTable/Calculation/PivotTableColumnCalculation.cs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		{
+			var calcOrder = GetCalcOrder();
+			foreach(var i in calcOrder)
+			{
+				var f = _tbl.Fields[i];
+				var tokens = f.Cache.FormulaTokens;
+				var calcTokens = GetPivotFieldReferencesInFormula(f, tokens);
+				if(calcTokens.Any(x => x == null))
+				{
+					//Contains invalid field reference or functions not supported in PT.
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/Calculation/PivotTableColumnCalculation.cs:55]: Is there a reason we don't throw or return an excel error when this happens?

File: src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+            {
+                currentKey[i] = PivotCalculationStore.SumLevelValue;
+                i++;
+                if (i == currentKey.Length) return false;
+				currentKey[i] = (currentKey[i] == PivotCalculationStore.SumLevelValue ? 0 : currentKey[i] + 1);
+			}
+			return true;
+        }
+
+        private bool IsSameLevelAs(int[] key, bool isRowField, int baseLevel, int keyCol, ExcelPivotTableDataField df)
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs]: Hides inherited member. Should have a 'new' keyword?

File: src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private int GetIndexPos(ExcelPivotTableField field)
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs]: No refences

File: src/EPPlus/FormulaParsing/DependencyChain/RpnFormulaExecution.cs (hunk 1 above)
Comment (JanKallman) [reply src/EPPlus/FormulaParsing/DependencyChain/RpnFormulaExecution.cs:803]: Looks ok in VS

File: src/EPPlus/Table/PivotTable/ExcelPivotTableFieldCollectionBase.cs [hunk 9]
```diff
@@ -118,8 +175,72 @@ public void SelectSingleItem(int index)
...
+                {
+					return 1;
+				}
+			}
+
+			private string GetTextValue(ExcelPivotTableFieldItem item)
+			{
+				if(string.IsNullOrEmpty(item.Text))
+                {
+                    //if(_hasGrouping)
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/ExcelPivotTableFieldCollectionBase.cs]: Remove comment?

File: src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs (hunk 2 above)
Comment (JanKallman) [reply src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs:184]: This adds the criteria field when having the string syntax in the GetPivotData function. This is by design

File: src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs [hunk 10]
```diff
@@ -378,6 +448,203 @@ internal void CreateValueFilter()
...
+        /// Handle date filters. This is filter enum values below 100.
+        /// </summary>
+        /// <param name="pivotTable"></param>
+        /// <param name="index"></param>
+        /// <returns></returns>
+        private bool MatchDate(ExcelPivotTable pivotTable, int index)
+        {
+            var date = ConvertUtil.GetValueDate(pivotTable.Fields[Fld].Cache.SharedItems[index]);
+            var value1Date = ConvertUtil.GetValueDate(Value1);
+            if (date.HasValue && value1Date.HasValue)
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs:521]: Switch case should have a default for "hängslen och livrem"

File: src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs [hunk 11]
```diff
@@ -378,6 +448,203 @@ internal void CreateValueFilter()
...
+            }
+            return false;
+        }
+        internal bool MatchNumeric(object value)
+        {
+            var num = RoundingHelper.RoundToSignificantFig(ConvertUtil.GetValueDouble(value), 15); 
+            var value1 = RoundingHelper.RoundToSignificantFig(ConvertUtil.GetValueDouble(Value1, false, true), 15);
+            if (double.IsNaN(num) == false && double.IsNaN(value1) == false)
+            {
+                switch (Type)
```
Comment (OssianEPPlus) [inline src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs:626]: This should also have a default

File: src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs (hunk 3 above)
Comment (JanKallman) [reply src/EPPlus/FormulaParsing/Excel/Functions/RefAndLookup/GetPivotData.cs:224]: The comments are incorrect. I'll change. 
Count - Count cells with numbers. 
CountA - Count cells that are not empty.

File: src/EPPlus/Table/PivotTable/Calculation/ArrayComparer.cs (hunk 5 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Calculation/ArrayComparer.cs:25]: This function is now used from three places where we created the keys dictionary.

File: src/EPPlus/Style/ExcelNumberFormat.cs (hunk 4 above)
Comment (JanKallman) [reply src/EPPlus/Style/ExcelNumberFormat.cs:209]: Should be left for future use.

File: src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs (hunk 8 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs]: Removed

File: src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs (hunk 7 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Calculation/ShowDataAs/PivotShowAsPercentOf.cs]: Removed. Base function is the same.

File: src/EPPlus/Table/PivotTable/Calculation/PivotTableColumnCalculation.cs (hunk 6 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Calculation/PivotTableColumnCalculation.cs:55]: Good point. I will throw an exception here, if the formula contains an invalid reference or unsupported function.

File: src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs (hunk 11 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs:626]: Unhandled enum values do not have a value (for example `LastMonth`), so default does nothing in this case.

File: src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs (hunk 10 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs:521]: Added default: with Exception

File: src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs (hunk 11 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/Filter/ExcelPivotTableFilter.cs:626]: Above comment refer to `switch` starting on row 60 (not commented)
Added exception on `switch` starting on row 624.

File: src/EPPlus/Table/PivotTable/ExcelPivotTableFieldCollectionBase.cs (hunk 9 above)
Comment (JanKallman) [reply src/EPPlus/Table/PivotTable/ExcelPivotTableFieldCollectionBase.cs]: Removed

**Pull request merged**